File: 2018/C/Day-2.cs
using System;
using System.Collections.Generic;

class Day2
{
    public int PartOne(string[] input)
    {
        int twos = 0;
        int threes = 0;

        foreach (string str in input)
        {
            Dictionary<char, int> counts = new Dictionary<char, int>();

            foreach (char letter in str)
            {
                int value;
                if (counts.TryGetValue(letter, out value))
                    counts[letter] = value + 1;
                else
                    counts[letter] = 1;
            }

            if (counts.ContainsValue(2))
                twos++;

            if (counts.ContainsValue(3))
                threes++;
        }

        return threes * twos;
    }

    public string PartTwo(string[] input)
    {
        foreach (string str in input)
        {
            foreach (string nextStr in input)
            {
                int numDifferences = 0;
                string same = "";

                for (int charIdx = 0; charIdx < str.Length; charIdx++)
                {
                    if (str[charIdx] != nextStr[charIdx])
                        numDifferences++;
                    else
                        same += str[charIdx];
                }

                if (numDifferences == 1)
                    return same;
            }
        }

        return "";
    }
}

File: 2018/Javascript/day-2.js
function part1(input) {
  let twos = 0;
  let threes = 0;

  for (let i = 0; i < input.length; i++) {
    const str = input[i];
    const counts = {};

    for (let charIdx = 0; charIdx < str.length; charIdx++) {
      const letter = str[charIdx];
      if (counts[letter]) {
        counts[letter]++;
      } else {
        counts[letter] = 1;
      }
    }

    const finalCounts = Object.values(counts);
    if (finalCounts.find(num => num === 2)) {
      twos++;
    }

    if (finalCounts.find(num => num === 3)) {
      threes++;
    }
  }

  return threes * twos;
}

function part2(input) {
  // iterate through every string.
  for (let strIdx = 0; strIdx < input.length; strIdx++) {
    const str = input[strIdx];

    // iterate through all the strings after it, starting with the next one
    for (let nextIdx = 0; nextIdx < str.length; nextIdx++) {
      let numDifferences = 0;
      const nextStr = input[nextIdx];

      // iterate through all the characters
      for (let charIdx = 0; charIdx < str.length; charIdx++) {
        if (str[charIdx] !== nextStr[charIdx]) {
          numDifferences++;
        }
      }

      if (numDifferences === 1) {
        // we found it...
        console.log(`FOUND 1: ${ str } 2: ${ nextStr }`);
        return;
      }
    }
  }
}
